### PR TITLE
Windows Support

### PR DESCRIPTION
**Pull request by 8ightfold**
You know what this is...

### CODE REVIEW

File: include/librepr/type/detail/undecorate.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+#include <Windows.h>
+#include <DbgHelp.h>
+
+#include <array>
+#include <mutex>
+#include <typeinfo>
+#include "denoise.h"
+
+// Awful!! But simplifies the _DecoratedName -> Symbol conversion
+#define M2B_CAST(b, m) [](auto* __ptr) { \
```
Comment (Tsche) [inline include/librepr/type/detail/undecorate.h:14]: does this need to be a macro?

File: include/librepr/type/detail/undecorate.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    // DON'T pass random strings starting with `.`
+    if (*symdata == '.') [[likely]] {
+      using uname_type = ::__std_type_info_data;
+      symdata = M2B_CAST(uname_type, _DecoratedName)(
+        symdata)->_UndecoratedName;
+      if(!symdata) [[unlikely]] { return false; }
+      auto denoised = denoise_name({ symdata });
+      std::memcpy(buf.data(), denoised.data(), denoised.size());
+      return denoised.size();
+    }
```
Comment (Tsche) [inline include/librepr/type/detail/undecorate.h:54]: should probably place a `std::unreachable();` here

File: include/librepr/type/name.h [hunk 3]
```diff
@@ -74,14 +75,27 @@ struct TemplateInfo<U<Ts...>> {
...
-#if defined(_WIN32)
-  return typeid(T).raw_name();
+#if USING(REPR_MSVC)
+  // Just in case...
+  if constexpr(detail::msvc::has_rawname) {
+    static auto& ty = typeid(T);
+    // .name() must be called since we have lazy symbol
+    // loading enabled. Otherwise when accessing later
+    // ._UndecoratedName will be NULL.
+    static volatile auto _ = ty.name();
```
Comment (Tsche) [inline include/librepr/type/name.h:89]: oof. Can you give it some name to stop clang-tidy from complaining though? :P

Minimum length is set to 3 characters I think

File: include/librepr/type_info.h [hunk 4]
```diff
@@ -24,18 +25,23 @@ struct TypeName {
 };
 }  // namespace librepr
 
-// std::format support for librepr::TypeInfo<T>{} and typeinfo<T>
+namespace librepr {
+template <class CharT>
+using StrFormatter = REPR_FORMAT_NS::formatter<std::string, CharT>;
```
Comment (Tsche) [inline include/librepr/type_info.h:30]: this should probably be in the `librepr::detail` namespace

**Review by Tsche — CHANGES_REQUESTED** [review]

File: include/librepr/type/detail/denoise.h [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+      return '\0';
+    }
+    return this->begin[pos];
+  }
+
+  REPR_MINLINE void write(SymToken tok) {
+    buf.write(to_sv(tok));
+  }
+
+  std::string_view to_sv(SymToken tok) const {
```
[the remark this answers is not in the document]
Comment (8ightfold) [reply include/librepr/type/detail/denoise.h:164]: ? Not sure what this means... I cant move this into the token class

File: include/librepr/type/detail/denoise.h [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    this->end = begin + sym.size();
+  }
+
+  static REPR_MINLINE bool isWord(char c) {
+    return (c >= 'A' && c <= 'Z') ||
+      (c >= 'a' && c <= 'z') ||
+      (c == '_');
+  }
+
+  static REPR_MINLINE bool isIdent(char c) {
```
[the remark this answers is not in the document]
Comment (8ightfold) [reply include/librepr/type/detail/denoise.h:113]: Yeah it was just 4am and I couldn't think of a name

File: include/librepr/type/detail/denoise.h [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+
+struct SymLexer {
+protected:
+  SymLexer(std::string_view sym) 
+   : begin(sym.data()), loc(begin),
+   buf(sym), nest_depth(), lcurr(), llast() {
+    this->end = begin + sym.size();
+  }
+
+  static REPR_MINLINE bool isWord(char c) {
```
[the remark this answers is not in the document]
Comment (8ightfold) [reply include/librepr/type/detail/denoise.h:107]: Word like regex's ``\w``

File: include/librepr/type/detail/denoise.h [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+    return (c >= 'A' && c <= 'Z') ||
```
[the remark this answers is not in the document]
Comment (8ightfold) [reply include/librepr/type/detail/denoise.h:108]: Nope, wouldn't be legal

File: include/librepr/type/detail/undecorate.h (hunk 1 above)
Comment (8ightfold) [reply include/librepr/type/detail/undecorate.h:14]: It's easier to follow imo

File: include/librepr/util/strings.h [hunk 9]
```diff
@@ -1,8 +1,31 @@
 #pragma once
 
+#include <concepts>
 #include <string>
 
+#define REPR_STRSIZE(...) ::librepr::detail::strsize(__VA_ARGS__)
```
[the remark this answers is not in the document]
Comment (8ightfold) [reply include/librepr/util/strings.h:6]: commas

File: include/librepr/type/detail/denoise.h [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+    }
+    return std::intptr_t(end - begin);
+  }
+
+private:
+  const char* begin;
+  const char* loc;
+  const char* end;
+  SymToken lcurr;
+  SymToken llast;
```
[the remark this answers is not in the document]
Comment (8ightfold) [reply include/librepr/type/detail/denoise.h:214]: lex

File: include/librepr/util/strings.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (Tsche) [reply include/librepr/util/strings.h:6]: wdym? Does this need to be a macro?

File: include/librepr/type/detail/denoise.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (Tsche) [reply include/librepr/type/detail/denoise.h:214]: Would `current_lexeme` and `last_lexeme` be more fitting then?

File: include/librepr/type/detail/denoise.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (Tsche) [reply include/librepr/type/detail/denoise.h:164]: Oh my bad, I think I misread this one.

File: include/librepr/type/detail/denoise.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (Tsche) [reply include/librepr/type/detail/denoise.h:107]: Okay, fair enough. I'd prefer snake case (`is_word`) though :)

File: include/librepr/type/detail/denoise.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (8ightfold) [reply include/librepr/type/detail/denoise.h:214]: Sure that would work

File: include/librepr/util/strings.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (8ightfold) [reply include/librepr/util/strings.h:6]: Probably not :P

**Review by Tsche — APPROVED** [review]

**Pull request merged**